# Business Understanding
## Overview of the business
A company that is seeking to expand into the aviation industry and wants buy aircrafts for its new division

## Business Problem

Being new in the aviation industry the company is unaware of the potential risks that are associated with aircrafts 

## Business Objectives
The company wants to understand the potential risks it may face in the aviation industry
The company wants to obtain a list of recommended aircrafts that are considered to have low risks in regards to its operational use

## Success Criteria
To obtain a list of aircrafts that are considered to have low risks

To outline the potential risks the company may experience in the aviation industry

To produce relevant visualisations that provide insights on aircraft and the aviation industry

## Project plan

To go through a dataset that contains accidents that have occurred in the aviation industry and to obtain relevant insights that can be able to solve the business problem






# Familization with the dataset

In [165]:
# Importing the pandas and matplotlib libraries
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import re
%matplotlib inline

In [112]:
pd.options.display.max_columns = None

Loading the dataset to be used


In [113]:
aviation_data_df = pd.read_csv("data/AviationData.csv", encoding='latin-1', dtype='str')


In [114]:
usstate_codes_df = pd.read_csv("data/USState_Codes.csv")

Checking the head, tail and a sample of the dataset so to get an understanding of the dataset we are working with


In [115]:
aviation_data_df.head()

Event.Id Investigation.Type Accident.Number  Event.Date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          Location        Country   Latitude   Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2    Saltville, VA  United States  36.922223  -81.878056          NaN   
3       EUREKA, CA  United States        NaN         NaN          NaN   
4       Canton, OH  United States        NaN         NaN          NaN   

  Airport.Name Injury.Severity Aircraft.damage Aircraft.Category  \
0          NaN        Fatal(2)       Destroyed               NaN   
1          NaN        Fatal(4)       Destroyed               NaN   
2          NaN        Fatal(3)       Destroyed               NaN   
3          NaN        Fatal(2)       Destroyed               NaN   
4          NaN        Fatal(1)       Destroyed               NaN   

  Registration.Number      Make     Model Amateur.Built Number.of.Engines  \
0              NC6404   Stinson     108-3            No                 1   
1              N5069P     Piper  PA24-180            No                 1   
2              N5142R    Cessna      172M            No                 1   
3              N1168J  Rockwell       112            No                 1   
4               N15NY    Cessna       501            No               NaN   

     Engine.Type FAR.Description Schedule Purpose.of.flight Air.carrier  \
0  Reciprocating             NaN      NaN          Personal         NaN   
1  Reciprocating             NaN      NaN          Personal         NaN   
2  Reciprocating             NaN      NaN          Personal         NaN   
3  Reciprocating             NaN      NaN          Personal         NaN   
4            NaN             NaN      NaN          Personal         NaN   

  Total.Fatal.Injuries Total.Serious.Injuries Total.Minor.Injuries  \
0                    2                      0                    0   
1                    4                      0                    0   
2                    3                    NaN                  NaN   
3                    2                      0                    0   
4                    1                      2                  NaN   

  Total.Uninjured Weather.Condition Broad.phase.of.flight   Report.Status  \
0               0               UNK                Cruise  Probable Cause   
1               0               UNK               Unknown  Probable Cause   
2             NaN               IMC                Cruise  Probable Cause   
3               0               IMC                Cruise  Probable Cause   
4               0               VMC              Approach  Probable Cause   

  Publication.Date  
0              NaN  
1       19-09-1996  
2       26-02-2007  
3       12-09-2000  
4       16-04-1980

In [116]:
aviation_data_df.tail()

Event.Id Investigation.Type Accident.Number  Event.Date  \
88884  20221227106491           Accident      ERA23LA093  2022-12-26   
88885  20221227106494           Accident      ERA23LA095  2022-12-26   
88886  20221227106497           Accident      WPR23LA075  2022-12-26   
88887  20221227106498           Accident      WPR23LA076  2022-12-26   
88888  20221230106513           Accident      ERA23LA097  2022-12-29   

            Location        Country Latitude Longitude Airport.Code  \
88884  Annapolis, MD  United States      NaN       NaN          NaN   
88885    Hampton, NH  United States      NaN       NaN          NaN   
88886     Payson, AZ  United States  341525N  1112021W          PAN   
88887     Morgan, UT  United States      NaN       NaN          NaN   
88888     Athens, GA  United States      NaN       NaN          NaN   

      Airport.Name Injury.Severity Aircraft.damage Aircraft.Category  \
88884          NaN           Minor             NaN               NaN   
88885          NaN             NaN             NaN               NaN   
88886       PAYSON       Non-Fatal     Substantial          Airplane   
88887          NaN             NaN             NaN               NaN   
88888          NaN           Minor             NaN               NaN   

      Registration.Number                        Make      Model  \
88884              N1867H                       PIPER  PA-28-151   
88885              N2895Z                    BELLANCA       7ECA   
88886              N749PJ  AMERICAN CHAMPION AIRCRAFT      8GCBC   
88887              N210CU                      CESSNA       210N   
88888              N9026P                       PIPER  PA-24-260   

      Amateur.Built Number.of.Engines Engine.Type FAR.Description Schedule  \
88884            No               NaN         NaN             091      NaN   
88885            No               NaN         NaN             NaN      NaN   
88886            No                 1         NaN             091      NaN   
88887            No               NaN         NaN             091      NaN   
88888            No               NaN         NaN             091      NaN   

      Purpose.of.flight         Air.carrier Total.Fatal.Injuries  \
88884          Personal                 NaN                    0   
88885               NaN                 NaN                    0   
88886          Personal                 NaN                    0   
88887          Personal  MC CESSNA 210N LLC                    0   
88888          Personal                 NaN                    0   

      Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
88884                      1                    0               0   
88885                      0                    0               0   
88886                      0                    0               1   
88887                      0                    0               0   
88888                      1                    0               1   

      Weather.Condition Broad.phase.of.flight Report.Status Publication.Date  
88884               NaN                   NaN           NaN       29-12-2022  
88885               NaN                   NaN           NaN              NaN  
88886               VMC                   NaN           NaN       27-12-2022  
88887               NaN                   NaN           NaN              NaN  
88888               NaN                   NaN           NaN       30-12-2022

In [117]:
aviation_data_df.sample(10)

Event.Id Investigation.Type Accident.Number  Event.Date  \
62676  20070712X00928           Accident      SEA07CA149  2007-06-10   
28957  20001212X17919           Accident      MIA91FA208  1991-08-28   
6305   20001214X44650           Accident      MKC83LA219  1983-09-18   
77352  20150923X04608           Accident      GAA15CA270  2015-09-21   
35153  20001206X01221           Accident      BFO94FA082  1994-05-28   
2284   20020917X03689           Accident      LAX82FA291  1982-08-08   
33140  20001211X12979           Accident      MIA93LA148  1993-07-02   
29077  20001212X18097           Accident      FTW91LA161  1991-09-14   
49390  20001212X21956           Accident      DEN00LA172  2000-09-20   
65625  20081112X70016           Accident      ERA09LA051  2008-11-12   

              Location        Country   Latitude  Longitude Airport.Code  \
62676      ESPARTO, CA  United States  38.650278  -122.0025         NONE   
28957        MIAMI, FL  United States        NaN        NaN          MIA   
6305        BENTON, KS  United States        NaN        NaN          1K1   
77352      Jackpot, NV  United States    415835N   1143927W          06U   
35153  LEONARDTOWN, MD  United States        NaN        NaN          NaN   
2284   WEAVERVILLE, CA  United States        NaN        NaN          054   
33140        TAMPA, FL  United States        NaN        NaN          NaN   
29077      BAYTOWN, TX  United States        NaN        NaN          NaN   
49390    ENGLEWOOD, CO  United States        NaN        NaN          APA   
65625     San Juan, PR  United States    182714N   0066532W         TJIG   

                        Airport.Name Injury.Severity Aircraft.damage  \
62676               Private Airstrip       Non-Fatal     Substantial   
28957            MIAMI INTERNATIONAL       Non-Fatal     Substantial   
6305                          BENTON       Non-Fatal       Destroyed   
77352           JACKPOT/HAYDEN FIELD       Non-Fatal     Substantial   
35153                            NaN       Non-Fatal       Destroyed   
2284                     WEAVERVILLE        Fatal(1)       Destroyed   
33140                            NaN       Non-Fatal       Destroyed   
29077                            NaN       Non-Fatal             NaN   
49390                     CENTENNIAL       Non-Fatal     Substantial   
65625  Fernando Luis Ribas Dominicci       Non-Fatal     Substantial   

      Aircraft.Category Registration.Number                 Make      Model  \
62676          Airplane              N5641H                Piper      PA-16   
28957               NaN              N41306                Piper   PA28-151   
6305                NaN              N7413V               Mooney       M20F   
77352          Airplane              N3354W                PIPER  PA 32-260   
35153               NaN              N5964J               Cessna      A150L   
2284           Airplane              N48754               Cessna        152   
33140               NaN              N4106M                Piper      PA-12   
29077               NaN              N5182U               Barnes       AX 7   
49390               NaN              N64796      Boeing Stearman      D75N1   
65625        Helicopter              N452JM  ROBINSON HELICOPTER        R44   

      Amateur.Built Number.of.Engines    Engine.Type  \
62676            No                 1  Reciprocating   
28957            No                 1  Reciprocating   
6305             No                 1  Reciprocating   
77352            No                 1  Reciprocating   
35153            No                 1  Reciprocating   
2284             No                 1  Reciprocating   
33140            No                 1  Reciprocating   
29077            No                 0        Unknown   
49390            No                 1  Reciprocating   
65625            No                 1  Reciprocating   

                 FAR.Description Schedule Purpose.of.flight       Air.carrier  \
62676  Part 91: Genera

In [118]:
aviation_data_df.columns

Index(['Event.Id', 'Investigation.Type', 'Accident.Number', 'Event.Date',
       'Location', 'Country', 'Latitude', 'Longitude', 'Airport.Code',
       'Airport.Name', 'Injury.Severity', 'Aircraft.damage',
       'Aircraft.Category', 'Registration.Number', 'Make', 'Model',
       'Amateur.Built', 'Number.of.Engines', 'Engine.Type', 'FAR.Description',
       'Schedule', 'Purpose.of.flight', 'Air.carrier', 'Total.Fatal.Injuries',
       'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured',
       'Weather.Condition', 'Broad.phase.of.flight', 'Report.Status',
       'Publication.Date'],
      dtype='object')

In [119]:
aviation_data_df.shape

(88889, 31)

In [120]:
len(aviation_data_df.columns)

31

In [121]:
usstate_codes_df.head()

US_State Abbreviation
0     Alabama           AL
1      Alaska           AK
2     Arizona           AZ
3    Arkansas           AR
4  California           CA

In [122]:
usstate_codes_df.tail()

US_State Abbreviation
57  Virgin Islands           VI
58   Washington_DC           DC
59  Gulf of mexico           GM
60  Atlantic ocean           AO
61   Pacific ocean           PO

In [123]:
usstate_codes_df.shape

(62, 2)

In [124]:
usstate_codes_df.columns

Index(['US_State', 'Abbreviation'], dtype='object')

# DATA UNDERSTANDING



The data folder contains two data sets

## Aviation dataset
This dataset contains data of reported aviation accidents and information regarding the accident reported.

The dataset contains 31 columns with each column containing information on the accident reported.

Some of the columns in the dataset include:date of the accident, the severity of the accident, country where the accident occurred, aircraft damage, aircraft type and model etc

The dataset contains 88889 records

## US State Codes dataset
This dataset contains information on the full name of abbreviated us state codes

The dataset contains two columns i.e the US state and its abbrevaiation

The dataset contains 62 records





# Data cleaning

## Aviation Dataset

In [125]:
aviation_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Event.Id                88889 non-null  object
 1   Investigation.Type      88889 non-null  object
 2   Accident.Number         88889 non-null  object
 3   Event.Date              88889 non-null  object
 4   Location                88837 non-null  object
 5   Country                 88663 non-null  object
 6   Latitude                34382 non-null  object
 7   Longitude               34373 non-null  object
 8   Airport.Code            50249 non-null  object
 9   Airport.Name            52790 non-null  object
 10  Injury.Severity         87889 non-null  object
 11  Aircraft.damage         85695 non-null  object
 12  Aircraft.Category       32287 non-null  object
 13  Registration.Number     87572 non-null  object
 14  Make                    88826 non-null  object
 15  Mo

Using pd.info we can observe the following

a) every column contain string values

b)They are columns with null values

c)Columns names contain different structures and need to be standardized

### Converting column names to a simple standardized form

a)column names to be converted to lower case

b)column names to be seperated by underscore


In [126]:
#converting the columns into lower case
aviation_data_df.columns=aviation_data_df.columns.str.lower()

In [127]:
#seperating column names with an underscore rather than with a period

aviation_data_df.columns=aviation_data_df.columns.str.replace('.','_')

aviation_data_df.columns

Index(['event_id', 'investigation_type', 'accident_number', 'event_date',
       'location', 'country', 'latitude', 'longitude', 'airport_code',
       'airport_name', 'injury_severity', 'aircraft_damage',
       'aircraft_category', 'registration_number', 'make', 'model',
       'amateur_built', 'number_of_engines', 'engine_type', 'far_description',
       'schedule', 'purpose_of_flight', 'air_carrier', 'total_fatal_injuries',
       'total_serious_injuries', 'total_minor_injuries', 'total_uninjured',
       'weather_condition', 'broad_phase_of_flight', 'report_status',
       'publication_date'],
      dtype='object')

In [128]:
aviation_data_df.head()

event_id investigation_type accident_number  event_date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          location        country   latitude   longitude airport_code  \
0  MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2    Saltville, VA  United States  36.922223  -81.878056          NaN   
3       EUREKA, CA  United States        NaN         NaN          NaN   
4       Canton, OH  United States        NaN         NaN          NaN   

  airport_name injury_severity aircraft_damage aircraft_category  \
0          NaN        Fatal(2)       Destroyed               NaN   
1          NaN        Fatal(4)       Destroyed               NaN   
2          NaN        Fatal(3)       Destroyed               NaN   
3          NaN        Fatal(2)       Destroyed               NaN   
4          NaN        Fatal(1)       Destroyed               NaN   

  registration_number      make     model amateur_built number_of_engines  \
0              NC6404   Stinson     108-3            No                 1   
1              N5069P     Piper  PA24-180            No                 1   
2              N5142R    Cessna      172M            No                 1   
3              N1168J  Rockwell       112            No                 1   
4               N15NY    Cessna       501            No               NaN   

     engine_type far_description schedule purpose_of_flight air_carrier  \
0  Reciprocating             NaN      NaN          Personal         NaN   
1  Reciprocating             NaN      NaN          Personal         NaN   
2  Reciprocating             NaN      NaN          Personal         NaN   
3  Reciprocating             NaN      NaN          Personal         NaN   
4            NaN             NaN      NaN          Personal         NaN   

  total_fatal_injuries total_serious_injuries total_minor_injuries  \
0                    2                      0                    0   
1                    4                      0                    0   
2                    3                    NaN                  NaN   
3                    2                      0                    0   
4                    1                      2                  NaN   

  total_uninjured weather_condition broad_phase_of_flight   report_status  \
0               0               UNK                Cruise  Probable Cause   
1               0               UNK               Unknown  Probable Cause   
2             NaN               IMC                Cruise  Probable Cause   
3               0               IMC                Cruise  Probable Cause   
4               0               VMC              Approach  Probable Cause   

  publication_date  
0              NaN  
1       19-09-1996  
2       26-02-2007  
3       12-09-2000  
4       16-04-1980

### Converting columns that contain string values to respective dtypes

This columns were all converted to string values while reading the csv so to clear error generated.Therefore we need to go through each column and with intuition convert some of the columns back to their original data type

In [129]:
aviation_data_df.sample(10)

event_id investigation_type accident_number  event_date  \
49499  20001212X22188           Accident      LAX01LA008  2000-10-07   
5052   20001214X43469           Accident      MKC83LA129  1983-06-15   
76193  20150325X90423           Incident      DCA15WA094  2015-01-13   
40291  20001208X06425           Accident      CHI96FA284  1996-08-08   
17901  20001213X31291           Accident      MKC87LA129  1987-06-27   
28911  20001212X17974           Accident      SEA91FA220  1991-08-23   
62126  20070314X00295           Accident      LAX07LA099  2007-03-03   
8815   20001214X40087           Accident      NYC84FA227  1984-06-30   
48464  20001212X21027           Accident      LAX00FA207  2000-05-27   
56420  20040402X00414           Accident      NYC04WA098  2004-03-30   

                         location         country   latitude    longitude  \
49499                    HILO, HI   United States        NaN          NaN   
5052                   PUXICO, MO   United States        NaN          NaN   
76193        Hong Kong, Hong Kong       Hong Kong        NaN          NaN   
40291                  MUNCIE, IN   United States        NaN          NaN   
17901             BULL SHOALS, AR   United States        NaN          NaN   
28911               ARLINGTON, WY   United States        NaN          NaN   
62126                   Arvin, CA   United States  35.223334  -118.841944   
8815                   BOSTON, MA   United States        NaN          NaN   
48464                PRESCOTT, AZ   United States        NaN          NaN   
56420  LANCASHIRE, UNITED KINGDOM  United Kingdom  53.865833    -1.660278   

      airport_code             airport_name injury_severity aircraft_damage  \
49499          ITO       HILO INTERNATIONAL       Non-Fatal     Substantial   
5052           NaN                      NaN       Non-Fatal       Destroyed   
76193          NaN                      NaN       Non-Fatal             NaN   
40291          MIE      DELAWARE CO JOHNSON        Fatal(2)       Destroyed   
17901          NaN                      NaN       Non-Fatal     Substantial   
28911          NaN                      NaN        Fatal(3)       Destroyed   
62126          NaN                      NaN        Fatal(1)       Destroyed   
8815           BOS  GENERAL EDWARD LAWRENCE        Fatal(1)       Destroyed   
48464          NaN                      NaN        Fatal(2)       Destroyed   
56420          NaN                      NaN        Fatal(1)     Substantial   

      aircraft_category registration_number        make         model  \
49499               NaN              N3715R      Cessna          172H   
5052                NaN              N4813Y       Piper     PA-25-235   
76193          Airplane               B8256  GULFSTREAM  GV - SP G550   
40291               NaN              N9137N       Piper     PA-28-181   
17901               NaN              N9161C      Cessna           180   
28911               NaN              N14698    Bellanca        17-30A   
62126               NaN              N789JM     Mallian     Rocket II   
8815                NaN              N120PB      Cessna          402C   
48464               NaN              N5207U      Cessna         172RG   
56420               NaN              N6834L      Cessna         T310R   

      amateur_built number_of_engines    engine_type far_description schedule  \
49499            No                 1  Reciprocating             NaN      NaN   
5052             No                 1  Reciprocating             NaN      UNK   
76193            No               NaN            NaN             NaN      NaN   
40291            No                 1  Reciprocating             NaN      NaN   
17901            No                 1  Reciprocating             NaN      NaN   
28911            No                 1  Reciprocating             NaN      NaN   
62126           Yes                 1  Reciprocating             NaN      NaN   
8815             No                 2  Reciprocati

#### number_of_engines,total_fatal_injuries,total_minor_injuries,total_serious_injuries and total_uninjured to interger values
Taking a sample of the dataset we can safely deduce that the following columns originally contained interger values

number_of_engines,total_fatal_injuries,total_minor_injuries,total_serious_injuries and total_uninjured

In [130]:
#We first convert the columns above to numeric
aviation_data_df['number_of_engines'] = pd.to_numeric(aviation_data_df['number_of_engines'],errors= 'coerce')
aviation_data_df['total_fatal_injuries'] = pd.to_numeric(aviation_data_df['total_fatal_injuries'], errors = 'coerce')
aviation_data_df['total_serious_injuries'] = pd.to_numeric(aviation_data_df['total_serious_injuries'], errors = 'coerce')
aviation_data_df['total_minor_injuries'] = pd.to_numeric(aviation_data_df['total_minor_injuries'], errors = 'coerce')
aviation_data_df['total_uninjured'] = pd.to_numeric(aviation_data_df['total_uninjured'], errors = 'coerce')


In [131]:

aviation_data_df[['number_of_engines','total_fatal_injuries','total_serious_injuries','total_minor_injuries','total_uninjured']]=aviation_data_df[['number_of_engines','total_fatal_injuries','total_serious_injuries','total_minor_injuries','total_uninjured']].astype('Int32')

In [132]:
aviation_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   event_id                88889 non-null  object
 1   investigation_type      88889 non-null  object
 2   accident_number         88889 non-null  object
 3   event_date              88889 non-null  object
 4   location                88837 non-null  object
 5   country                 88663 non-null  object
 6   latitude                34382 non-null  object
 7   longitude               34373 non-null  object
 8   airport_code            50249 non-null  object
 9   airport_name            52790 non-null  object
 10  injury_severity         87889 non-null  object
 11  aircraft_damage         85695 non-null  object
 12  aircraft_category       32287 non-null  object
 13  registration_number     87572 non-null  object
 14  make                    88826 non-null  object
 15  mo

In [133]:
aviation_data_df.head()

event_id investigation_type accident_number  event_date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          location        country   latitude   longitude airport_code  \
0  MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2    Saltville, VA  United States  36.922223  -81.878056          NaN   
3       EUREKA, CA  United States        NaN         NaN          NaN   
4       Canton, OH  United States        NaN         NaN          NaN   

  airport_name injury_severity aircraft_damage aircraft_category  \
0          NaN        Fatal(2)       Destroyed               NaN   
1          NaN        Fatal(4)       Destroyed               NaN   
2          NaN        Fatal(3)       Destroyed               NaN   
3          NaN        Fatal(2)       Destroyed               NaN   
4          NaN        Fatal(1)       Destroyed               NaN   

  registration_number      make     model amateur_built  number_of_engines  \
0              NC6404   Stinson     108-3            No                  1   
1              N5069P     Piper  PA24-180            No                  1   
2              N5142R    Cessna      172M            No                  1   
3              N1168J  Rockwell       112            No                  1   
4               N15NY    Cessna       501            No               <NA>   

     engine_type far_description schedule purpose_of_flight air_carrier  \
0  Reciprocating             NaN      NaN          Personal         NaN   
1  Reciprocating             NaN      NaN          Personal         NaN   
2  Reciprocating             NaN      NaN          Personal         NaN   
3  Reciprocating             NaN      NaN          Personal         NaN   
4            NaN             NaN      NaN          Personal         NaN   

   total_fatal_injuries  total_serious_injuries  total_minor_injuries  \
0                     2                       0                     0   
1                     4                       0                     0   
2                     3                    <NA>                  <NA>   
3                     2                       0                     0   
4                     1                       2                  <NA>   

   total_uninjured weather_condition broad_phase_of_flight   report_status  \
0                0               UNK                Cruise  Probable Cause   
1                0               UNK               Unknown  Probable Cause   
2             <NA>               IMC                Cruise  Probable Cause   
3                0               IMC                Cruise  Probable Cause   
4                0               VMC              Approach  Probable Cause   

  publication_date  
0              NaN  
1       19-09-1996  
2       26-02-2007  
3       12-09-2000  
4       16-04-1980

#### converting event_date and publication_date to pandas.Datetime

In [134]:
#Converting columns that contain dates into pandas.datetime
# that is event_date and publication_date


aviation_data_df['event_date']=pd.to_datetime(aviation_data_df['event_date'],format="%Y-%m-%d" ,yearfirst= True, errors= 'coerce')



In [135]:
aviation_data_df['publication_date']=pd.to_datetime(aviation_data_df['publication_date'],format="%d-%m-%Y", yearfirst= True, errors='coerce')

In [136]:
aviation_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   event_id                88889 non-null  object        
 1   investigation_type      88889 non-null  object        
 2   accident_number         88889 non-null  object        
 3   event_date              88889 non-null  datetime64[ns]
 4   location                88837 non-null  object        
 5   country                 88663 non-null  object        
 6   latitude                34382 non-null  object        
 7   longitude               34373 non-null  object        
 8   airport_code            50249 non-null  object        
 9   airport_name            52790 non-null  object        
 10  injury_severity         87889 non-null  object        
 11  aircraft_damage         85695 non-null  object        
 12  aircraft_category       32287 non-null  object

In [137]:
aviation_data_df[['event_date','publication_date']].sample(20)

event_date publication_date
77098 2015-08-07       2020-09-25
17542 1987-05-18       1988-12-27
135   1982-01-20       1983-01-20
75189 2014-06-05       2020-09-25
31198 1992-08-16       1996-08-05
6072  1983-08-29              NaT
23600 1989-07-27       1991-09-30
48286 2000-05-01       2001-05-09
21028 1988-08-01       1990-10-16
35913 1994-08-28       1995-05-04
66502 2009-06-07       2020-09-25
17067 1987-03-21       1989-01-25
88168 2022-07-16              NaT
23610 1989-07-28       1991-02-15
78515 2016-06-29       2020-09-25
26507 1990-08-31       1992-12-15
81350 2018-04-05       2021-05-20
27577 1991-03-17       1993-03-31
80254 2017-07-15       2020-09-25
49888 2000-12-29       2003-08-07

In [138]:
#It might be useful to extract the years and the months of the event_date and publication_date which can be useful when doing visualisations
aviation_data_df['event_year'] = aviation_data_df['event_date'].dt.year
aviation_data_df['event_month'] = aviation_data_df['event_date'].dt.month_name()
aviation_data_df['publication_year'] = aviation_data_df['publication_date'].dt.year
aviation_data_df['publication_month'] = aviation_data_df['publication_date'].dt.month_name()

In [139]:
aviation_data_df.head()

event_id investigation_type accident_number event_date  \
0  20001218X45444           Accident      SEA87LA080 1948-10-24   
1  20001218X45447           Accident      LAX94LA336 1962-07-19   
2  20061025X01555           Accident      NYC07LA005 1974-08-30   
3  20001218X45448           Accident      LAX96LA321 1977-06-19   
4  20041105X01764           Accident      CHI79FA064 1979-08-02   

          location        country   latitude   longitude airport_code  \
0  MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2    Saltville, VA  United States  36.922223  -81.878056          NaN   
3       EUREKA, CA  United States        NaN         NaN          NaN   
4       Canton, OH  United States        NaN         NaN          NaN   

  airport_name injury_severity aircraft_damage aircraft_category  \
0          NaN        Fatal(2)       Destroyed               NaN   
1          NaN        Fatal(4)       Destroyed               NaN   
2          NaN        Fatal(3)       Destroyed               NaN   
3          NaN        Fatal(2)       Destroyed               NaN   
4          NaN        Fatal(1)       Destroyed               NaN   

  registration_number      make     model amateur_built  number_of_engines  \
0              NC6404   Stinson     108-3            No                  1   
1              N5069P     Piper  PA24-180            No                  1   
2              N5142R    Cessna      172M            No                  1   
3              N1168J  Rockwell       112            No                  1   
4               N15NY    Cessna       501            No               <NA>   

     engine_type far_description schedule purpose_of_flight air_carrier  \
0  Reciprocating             NaN      NaN          Personal         NaN   
1  Reciprocating             NaN      NaN          Personal         NaN   
2  Reciprocating             NaN      NaN          Personal         NaN   
3  Reciprocating             NaN      NaN          Personal         NaN   
4            NaN             NaN      NaN          Personal         NaN   

   total_fatal_injuries  total_serious_injuries  total_minor_injuries  \
0                     2                       0                     0   
1                     4                       0                     0   
2                     3                    <NA>                  <NA>   
3                     2                       0                     0   
4                     1                       2                  <NA>   

   total_uninjured weather_condition broad_phase_of_flight   report_status  \
0                0               UNK                Cruise  Probable Cause   
1                0               UNK               Unknown  Probable Cause   
2             <NA>               IMC                Cruise  Probable Cause   
3                0               IMC                Cruise  Probable Cause   
4                0               VMC              Approach  Probable Cause   

  publication_date  event_year event_month  publication_year publication_month  
0              NaT        1948     October               NaN               NaN  
1       1996-09-19        1962        July            1996.0         September  
2       2007-02-26        1974      August            2007.0          February  
3       2000-09-12        1977        June            2000.0         September  
4       1980-04-16        1979      August            1980.0             April

In [140]:
aviation_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 35 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   event_id                88889 non-null  object        
 1   investigation_type      88889 non-null  object        
 2   accident_number         88889 non-null  object        
 3   event_date              88889 non-null  datetime64[ns]
 4   location                88837 non-null  object        
 5   country                 88663 non-null  object        
 6   latitude                34382 non-null  object        
 7   longitude               34373 non-null  object        
 8   airport_code            50249 non-null  object        
 9   airport_name            52790 non-null  object        
 10  injury_severity         87889 non-null  object        
 11  aircraft_damage         85695 non-null  object        
 12  aircraft_category       32287 non-null  object

In [141]:
#due to the null values present in the publication year we need to convert the publication year to interger

aviation_data_df['publication_year'] = aviation_data_df['publication_year'].astype('Int64')

In [142]:
aviation_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 35 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   event_id                88889 non-null  object        
 1   investigation_type      88889 non-null  object        
 2   accident_number         88889 non-null  object        
 3   event_date              88889 non-null  datetime64[ns]
 4   location                88837 non-null  object        
 5   country                 88663 non-null  object        
 6   latitude                34382 non-null  object        
 7   longitude               34373 non-null  object        
 8   airport_code            50249 non-null  object        
 9   airport_name            52790 non-null  object        
 10  injury_severity         87889 non-null  object        
 11  aircraft_damage         85695 non-null  object        
 12  aircraft_category       32287 non-null  object

In [143]:
#Let us reorder our new created columns to be closer to their referenced column
column_names=aviation_data_df.columns.to_list()
column_names

['event_id',
 'investigation_type',
 'accident_number',
 'event_date',
 'location',
 'country',
 'latitude',
 'longitude',
 'airport_code',
 'airport_name',
 'injury_severity',
 'aircraft_damage',
 'aircraft_category',
 'registration_number',
 'make',
 'model',
 'amateur_built',
 'number_of_engines',
 'engine_type',
 'far_description',
 'schedule',
 'purpose_of_flight',
 'air_carrier',
 'total_fatal_injuries',
 'total_serious_injuries',
 'total_minor_injuries',
 'total_uninjured',
 'weather_condition',
 'broad_phase_of_flight',
 'report_status',
 'publication_date',
 'event_year',
 'event_month',
 'publication_year',
 'publication_month']

In [144]:
reorderd_columns=['event_id',
 'investigation_type',
 'accident_number',
 'event_date',
 'event_year',
 'event_month',
 'location',
 'country',
 'latitude',
 'longitude',
 'airport_code',
 'airport_name',
 'injury_severity',
 'aircraft_damage',
 'aircraft_category',
 'registration_number',
 'make',
 'model',
 'amateur_built',
 'number_of_engines',
 'engine_type',
 'far_description',
 'schedule',
 'purpose_of_flight',
 'air_carrier',
 'total_fatal_injuries',
 'total_serious_injuries',
 'total_minor_injuries',
 'total_uninjured',
 'weather_condition',
 'broad_phase_of_flight',
 'report_status',
 'publication_date',
 'publication_year',
 'publication_month']

In [145]:
aviation_data_df=aviation_data_df[reorderd_columns]

In [146]:
aviation_data_df.head()

event_id investigation_type accident_number event_date  event_year  \
0  20001218X45444           Accident      SEA87LA080 1948-10-24        1948   
1  20001218X45447           Accident      LAX94LA336 1962-07-19        1962   
2  20061025X01555           Accident      NYC07LA005 1974-08-30        1974   
3  20001218X45448           Accident      LAX96LA321 1977-06-19        1977   
4  20041105X01764           Accident      CHI79FA064 1979-08-02        1979   

  event_month         location        country   latitude   longitude  \
0     October  MOOSE CREEK, ID  United States        NaN         NaN   
1        July   BRIDGEPORT, CA  United States        NaN         NaN   
2      August    Saltville, VA  United States  36.922223  -81.878056   
3        June       EUREKA, CA  United States        NaN         NaN   
4      August       Canton, OH  United States        NaN         NaN   

  airport_code airport_name injury_severity aircraft_damage aircraft_category  \
0          NaN          NaN        Fatal(2)       Destroyed               NaN   
1          NaN          NaN        Fatal(4)       Destroyed               NaN   
2          NaN          NaN        Fatal(3)       Destroyed               NaN   
3          NaN          NaN        Fatal(2)       Destroyed               NaN   
4          NaN          NaN        Fatal(1)       Destroyed               NaN   

  registration_number      make     model amateur_built  number_of_engines  \
0              NC6404   Stinson     108-3            No                  1   
1              N5069P     Piper  PA24-180            No                  1   
2              N5142R    Cessna      172M            No                  1   
3              N1168J  Rockwell       112            No                  1   
4               N15NY    Cessna       501            No               <NA>   

     engine_type far_description schedule purpose_of_flight air_carrier  \
0  Reciprocating             NaN      NaN          Personal         NaN   
1  Reciprocating             NaN      NaN          Personal         NaN   
2  Reciprocating             NaN      NaN          Personal         NaN   
3  Reciprocating             NaN      NaN          Personal         NaN   
4            NaN             NaN      NaN          Personal         NaN   

   total_fatal_injuries  total_serious_injuries  total_minor_injuries  \
0                     2                       0                     0   
1                     4                       0                     0   
2                     3                    <NA>                  <NA>   
3                     2                       0                     0   
4                     1                       2                  <NA>   

   total_uninjured weather_condition broad_phase_of_flight   report_status  \
0                0               UNK                Cruise  Probable Cause   
1                0               UNK               Unknown  Probable Cause   
2             <NA>               IMC                Cruise  Probable Cause   
3                0               IMC                Cruise  Probable Cause   
4                0               VMC              Approach  Probable Cause   

  publication_date  publication_year publication_month  
0              NaT              <NA>               NaN  
1       1996-09-19              1996         September  
2       2007-02-26              2007          February  
3       2000-09-12              2000         September  
4       1980-04-16              1980             April

#### Location,Country,Latitudes and longitude

In [147]:
#Let us check the sum of null values in the latitude and longitude
aviation_data_df.isnull().sum()

event_id                      0
investigation_type            0
accident_number               0
event_date                    0
event_year                    0
event_month                   0
location                     52
country                     226
latitude                  54507
longitude                 54516
airport_code              38640
airport_name              36099
injury_severity            1000
aircraft_damage            3194
aircraft_category         56602
registration_number        1317
make                         63
model                        92
amateur_built               102
number_of_engines          6084
engine_type                7077
far_description           56866
schedule                  76307
purpose_of_flight          6192
air_carrier               72241
total_fatal_injuries      11401
total_serious_injuries    12510
total_minor_injuries      11933
total_uninjured            5912
weather_condition          4492
broad_phase_of_flight     27165
report_s

In [148]:
aviation_data_df[['location','country','latitude','longitude']].sample(30)

location         country   latitude   longitude
60700                    MILLBROOK, NY   United States  41.704444  -73.736111
21416                   GREENVILLE, ME   United States        NaN         NaN
27886                      MEMPHIS, TN   United States        NaN         NaN
45579                         CODY, WY   United States        NaN         NaN
78412       Manchester, United Kingdom  United Kingdom        NaN         NaN
84890                        Atoka, TN   United States    352448N    0893911W
71604                   San Manuel, AZ   United States    323811N    1103850W
11389                         ALMA, GA   United States        NaN         NaN
56875                   St. George, UT   United States  37.326111   -113.6225
64002                     Marathon, FL   United States    245229N    0811348W
30665                     WOODLAND, WA   United States        NaN         NaN
47594                       SALINA, KS   United States        NaN         NaN
23881                       HEARNE, TX   United States        NaN         NaN
77523       Johannesburg, South Africa    South Africa    002681S    0281432E
88689                    Rudraprayag,            India        NaN         NaN
48301                      EPHRATA, WA   United States        NaN         NaN
50171                      BOZEMAN, MT   United States        NaN         NaN
45661                      GRENOLA, KS   United States        NaN         NaN
86781               Maturin, Venezuela       Venezuela    094457N    0063912W
56253                   CANNES, FRANCE          France        NaN         NaN
6604                   BAKERSFIELD, CA   United States        NaN         NaN
24848                      PAHOKEE, FL   United States        NaN         NaN
87938                Fort Duchesne, UT   United States    004083N    1094727W
16222                  BLOOMINGTON, MD   United States        NaN         NaN
4104                     SOUTHSIDE, AL   United States        NaN         NaN
45914                    PACIFIC OCEAN   PACIFIC OCEAN        NaN         NaN
2686                       ANDOVER, KS   United States        NaN         NaN
49187                    CORSICANA, TX   United States        NaN         NaN
72302                       Denver, CO   United States    395142N    1044023W
77300  Santa Fe de Antioquia, Colombia        Colombia        NaN         NaN

In [149]:
#From the above analysis majority of the latitude and longitude values are null values
#We can generate the longitudes and latitudes of the different locations using values in the location and country columns

aviation_data_df[aviation_data_df['country'].isna()]

event_id investigation_type accident_number event_date  \
36     20020917X02410           Accident      MIA82FKA05 1982-01-04   
464    20020917X02358           Accident      MIA82DA062 1982-03-02   
465    20020917X02026           Accident      FTW82DA076 1982-03-02   
725    20020917X02377           Accident      MIA82DA091 1982-03-31   
831    20020917X02069           Accident      FTW82DA127 1982-04-13   
...               ...                ...             ...        ...   
52288  20020322X00387           Accident      DCA02MA029 2002-03-22   
54284  20040528X00699           Accident      DCA03WA031 2003-03-12   
56200  20040528X00697           Accident      DCA04WA026 2004-02-10   
58803  20050616X00790           Incident      DCA05WA073 2005-06-10   
62530  20070518X00582           Incident      DCA07WA043 2007-05-13   

       event_year event_month         location country latitude longitude  \
36           1982     January      SAINT CROIX     NaN      NaN       NaN   
464          1982       March         HUMA CAO     NaN      NaN       NaN   
465          1982       March  MUSTANG BLK A11     NaN      NaN       NaN   
725          1982       March             MOCA     NaN      NaN       NaN   
831          1982       April  WEST DELTA 105D     NaN      NaN       NaN   
...           ...         ...              ...     ...      ...       ...   
52288        2002       March     int'l waters     NaN      NaN       NaN   
54284        2003       March     JOHANNESBURG     NaN      NaN       NaN   
56200        2004    February  Sharjah Airport     NaN      NaN       NaN   
58803        2005        June  Canada/US borde     NaN      NaN       NaN   
62530        2007         May   London Control     NaN      NaN       NaN   

      airport_code        airport_name injury_severity aircraft_damage  \
36             STX  ALEXANDER HAMILTON        Fatal(1)       Destroyed   
464            NaN                 NaN        Fatal(1)       Destroyed   
465            NaN                 NaN       Non-Fatal             NaN   
725           PR05             LABADIE       Non-Fatal     Substantial   
831            NaN                 NaN       Non-Fatal       Destroyed   
...            ...                 ...             ...             ...   
52288          NaN                 NaN       Non-Fatal           Minor   
54284          NaN                 NaN     Unavailable     Substantial   
56200          NaN                 NaN       Fatal(46)       Destroyed   
58803          NaN                 NaN        Incident             NaN   
62530          NaN                 NaN        Incident             NaN   

      aircraft_category registration_number               make     model  \
36             Airplane              N5151U             Cessna       206   
464            Airplane              N2741J             Cessna       150   
465          Helicopter              N1080S               Bell        12   
725            Airplane              N309MJ              Piper    PA-32R   
831          Helicopter               N41AL               Bell      206B   
...                 ...                 ...                ...       ...   
52288               NaN              N234NW  Mcdonnell Douglas  DC-10-30   
54284               NaN                 NaN             Boeing  747-300F   
56200               NaN                 NaN             Fokker        50   
58803               NaN                 NaN                NaN       NaN   
62530               NaN                 NaN             Boeing   737-800   

      amateur_built  number_of_engines    engine_type  \
36               No                  1  Reciprocating   
464              No                  1  Reciprocating   
465              No                  2    Turbo Shaft   
725              No                  1  Reciprocating   
831              No                  1    Turbo Shaft   
...             ...                ...            ...   
52288            No               

In [150]:
# The other dataset contains the full abreviations of some of those states
usstate_codes_df

US_State Abbreviation
0          Alabama           AL
1           Alaska           AK
2          Arizona           AZ
3         Arkansas           AR
4       California           CA
..             ...          ...
57  Virgin Islands           VI
58   Washington_DC           DC
59  Gulf of mexico           GM
60  Atlantic ocean           AO
61   Pacific ocean           PO

[62 rows x 2 columns]

In [151]:
state_dict= dict(zip(usstate_codes_df['Abbreviation'],usstate_codes_df['US_State']))
state_dict

{'AL': 'Alabama',
 'AK': 'Alaska',
 'AZ': 'Arizona',
 'AR': 'Arkansas',
 'CA': 'California',
 'CO': 'Colorado',
 'CT': 'Connecticut',
 'DE': 'Delaware',
 'FL': 'Florida',
 'GA': 'Georgia',
 'HI': 'Hawaii',
 'ID': 'Idaho',
 'IL': 'Illinois',
 'IN': 'Indiana',
 'IA': 'Iowa',
 'KS': 'Kansas',
 'KY': 'Kentucky',
 'LA': 'Louisiana',
 'ME': 'Maine',
 'MD': 'Maryland',
 'MA': 'Massachusetts',
 'MI': 'Michigan',
 'MN': 'Minnesota',
 'MS': 'Mississippi',
 'MO': 'Missouri',
 'MT': 'Montana',
 'NE': 'Nebraska',
 'NV': 'Nevada',
 'NH': 'New Hampshire',
 'NJ': 'New Jersey',
 'NM': 'New Mexico',
 'NY': 'New York',
 'NC': 'North Carolina',
 'ND': 'North Dakota',
 'OH': 'Ohio',
 'OK': 'Oklahoma',
 'OR': 'Oregon',
 'PA': 'Pennsylvania',
 'RI': 'Rhode Island',
 'SC': 'South Carolina',
 'SD': 'South Dakota',
 'TN': 'Tennessee',
 'TX': 'Texas',
 'UT': 'Utah',
 'VT': 'Vermont',
 'VA': 'Virginia',
 'WA': 'Washington',
 'WV': 'West Virginia',
 'WI': 'Wisconsin',
 'WY': 'Wyoming',
 'AS': 'American Samoa',
 'G

In [152]:
#Lets replace the abbreviated state names with full names in the location column
aviation_data_df['location'] = aviation_data_df['location'].replace(state_dict,regex=True)


In [154]:
#Lets convert the locations into title cases
aviation_data_df['location'] = aviation_data_df['location'].str.title()
aviation_data_df.sample(20)

event_id investigation_type accident_number event_date  \
37258  20001207X03457           Accident      FTW95LA202 1995-05-14   
23692  20001213X29236           Accident      NYC89LA192 1989-08-05   
81321  20180423X51112           Incident      DCA18WA156 2018-03-27   
76562  20150520X84121           Incident      DCA15WA122 2015-04-28   
31481  20001211X15785           Accident      SEA92LA213 1992-09-19   
88358  20220818105759           Accident      GAA22WA255 2022-08-14   
62019  20070209X00166           Accident     LAX07LA079B 2007-02-03   
35955  20001206X02278           Accident      LAX94LA347 1994-09-02   
60113  20060130X00132           Accident      DFW06LA057 2006-01-19   
69786  20110411X31443           Accident      WPR11CA194 2011-04-08   
38921  20001208X05167           Accident      MIA96FA059 1996-01-07   
45330  20001211X11375           Accident      CHI99FA042 1998-11-29   
57640  20041025X01689           Accident      NYC05FA005 2004-10-11   
31351  20001211X15652           Accident      CHI92LA265 1992-09-04   
51529  20011004X02038           Accident      CHI01LA330 2001-09-26   
47581  20001212X20250           Accident      FTW00RA043 1999-12-11   
5174   20001214X43318           Accident      DEN83LTE11 1983-06-24   
77738  20160113X10829           Accident      WPR16WA052 2016-01-01   
77908  20160229X32454           Accident      GAA16CA138 2016-02-24   
18147  20001213X31611           Accident      MKC87LA142 1987-07-19   

       event_year event_month  \
37258        1995         May   
23692        1989      August   
81321        2018       March   
76562        2015       April   
31481        1992   September   
88358        2022      August   
62019        2007    February   
35955        1994   September   
60113        2006     January   
69786        2011       April   
38921        1996     January   
45330        1998    November   
57640        2004     October   
31351        1992   September   
51529        2001   September   
47581        1999    December   
5174         1983        June   
77738        2016     January   
77908        2016    February   
18147        1987        July   

                                                location        country  \
37258                                     Joes, Colorado  United States   
23692                    Grhode Islandswold, Connecticut  United States   
81321                                Mexico City, Mexico         Mexico   
76562                                   Barcelona, Spain          Spain   
31481                                     Lloyd, Montana  United States   
88358                                          Nipigon,          Canada   
62019                           Palos Verdes, California  United States   
35955                                   Mainesa, Arizona  United States   
60113                          Foregont Woregonth, Texas  United States   
69786                                Othello, Washington  United States   
38921  Namerican Samoahvirgin Islandsllinoisle, Tenne...  United States   
45330                Coloradolemassachusettsn, Wisconsin  United States   
57640                    Massachusettsdison, Connecticut  United States   
31351                              Missouriran, Michigan  United States   
51529                         Farkansasibault, Minnesota  United States   
47581                        Californialabamai, Colombia       Colombia   
5174                                    Granby, Colorado  United States   
77738                Seletar Airport, Singapore, Senegal        Senegal   
77908                               Enumclaw, Washington  United States   
18147                  Kansamerican Samoa City, Missouri  United States   

        latitude   longitude airport_code                    airport_name  \
37258        NaN         NaN          NaN                             NaN   
23692        NaN         NaN          5B4                        LAKESIDE   
81321        NaN    

In [164]:
aviation_data_df['location'].isna().sum()

52

In [177]:
# Changing 
def dms_to_dd(dms_str):
    # Regular expression to extract degrees, minutes, and seconds
    dms_pattern = re.compile(r'(\d{2,3})(\d{2})(\d{2})([NSEW])')
    match = dms_pattern.match(dms_str)
    if not match:
        print(f"No match for: {dms_str}")
        return None
    
    degrees, minutes, seconds, direction = match.groups()
    print(f"Matched: {degrees}° {minutes}' {seconds}\" {direction}")
    
    dd = float(degrees) + float(minutes)/60 + float(seconds)/3600
    
    if direction in ['S', 'W']:
        dd *= -1
    
    return dd

# Sample data
# df = pd.DataFrame({
#     'latitude': ['471146N', '36.7783', '407128N'],
#     'longitude': ['0122113W', '-119.4179', '0740060W']
# })
df = pd.DataFrame({'latitude':[],'longitude':[]})
# Apply the conversion function
df['latitude'] = aviation_data_df['latitude'].apply(lambda x: dms_to_dd(x) if re.match(r'\d{6}[NSEW]', x) else pd.to_numeric(x, errors='coerce'))
df['longitude'] = aviation_data_df['longitude'].apply(lambda x: dms_to_dd(x) if re.match(r'\d{7}[NSEW]', x) else pd.to_numeric(x, errors='coerce'))

print(df)






TypeError: expected string or bytes-like object

In [179]:

def dms_to_dd(dms_str):
    if pd.isna(dms_str):
        return np.nan
    
    # Regular expression to extract degrees, minutes, and seconds
    dms_pattern = re.compile(r'(\d{2,3})(\d{2})(\d{2})([NSEW])')
    match = dms_pattern.match(dms_str)
    if not match:
        print(f"No match for: {dms_str}")
        return np.nan
    
    degrees, minutes, seconds, direction = match.groups()
    print(f"Matched: {degrees}° {minutes}' {seconds}\" {direction}")
    
    dd = float(degrees) + float(minutes)/60 + float(seconds)/3600
    
    if direction in ['S', 'W']:
        dd *= -1
    
    return dd

# Sample data
df = pd.DataFrame({
    'latitude': [],
    'longitude': []
})

# Apply the conversion function
df['latitude'] = aviation_data_df['latitude'].apply(lambda x: dms_to_dd(x) if isinstance(x, str) and re.match(r'\d{6}[NSEW]', x) else pd.to_numeric(x, errors='coerce'))
df['longitude'] = aviation_data_df['longitude'].apply(lambda x: dms_to_dd(x) if isinstance(x, str) and re.match(r'\d{7}[NSEW]', x) else pd.to_numeric(x, errors='coerce'))

print(df)


Matched: 38° 13' 26" N
Matched: 29° 30' 22" N
Matched: 04° 02' 76" N
Matched: 00° 31' 96" N
Matched: 03° 53' 21" N
Matched: 48° 21' 48" N
Matched: 38° 58' 21" N
Matched: 42° 22' 35" N
Matched: 03° 81' 28" N
Matched: 05° 74' 52" N
Matched: 03° 94' 36" N
Matched: 28° 28' 24" N
Matched: 25° 37' 26" N
Matched: 41° 27' 12" N
Matched: 04° 43' 87" N
Matched: 02° 85' 94" N
Matched: 36° 14' 19" N
Matched: 04° 92' 40" N
Matched: 04° 05' 15" N
Matched: 33° 51' 18" N
Matched: 26° 49' 54" N
Matched: 42° 12' 45" N
Matched: 03° 35' 35" N
Matched: 04° 85' 10" N
Matched: 27° 15' 35" N
Matched: 22° 36' 42" S
Matched: 26° 46' 44" N
Matched: 04° 19' 12" N
Matched: 41° 31' 38" N
Matched: 46° 19' 17" N
Matched: 27° 54' 47" N
Matched: 45° 38' 55" N
Matched: 04° 01' 48" N
Matched: 38° 43' 27" N
Matched: 31° 58' 17" N
Matched: 45° 42' 59" N
Matched: 03° 55' 29" N
Matched: 02° 15' 12" N
Matched: 03° 73' 78" N
Matched: 03° 40' 57" N
Matched: 33° 52' 32" N
Matched: 60° 46' 47" N
Matched: 42° 59' 52" N
Matched: 61